### PR TITLE
Add background rendering to TextLayer

### PR DESCRIPTION
**Pull request by Pessimistress**
For #4579

This PR adds a "background" sublayer to the TextLayer. This allows for:

- Padding
- Per-object background color, border color, and border width
- (Advanced user) custom styling via `_subLayerProps` and custom layer

With `background: false`, the feature has no impact on perf. With `background: true` (as well as legacy `backgroundColor` users) there is more overhead in attribute generation, in exchange for the additional flexibility.

#### Change List
- Deprecate `backgroundColor` prop, add `background`, `backgroundPadding`, `getBackgroundColor`, `getBorderColor`, `getBorderWidth`
- Documentation
- Upgrade Guide
- Render test

### CODE REVIEW

File: modules/layers/src/text-layer/text-layer.js [hunk 1]
```diff
@@ -266,63 +304,121 @@ export default class TextLayer extends CompositeLayer {
...
-          getAngle: updateTriggers.getAngle,
-          getColor: updateTriggers.getColor,
-          getSize: updateTriggers.getSize,
-          getPixelOffset: updateTriggers.getPixelOffset,
-          getIconOffsets: {
-            getText: updateTriggers.getText,
-            getTextAnchor: updateTriggers.getTextAnchor,
-            getAlignmentBaseline: updateTriggers.getAlignmentBaseline,
-            styleVersion
+    const CharacgtersLayerClass = this.getSubLayerClass('characters', MultiIconLayer);
```
Comment (ibgreen) [inline modules/layers/src/text-layer/text-layer.js:307]: Typo

File: modules/layers/src/text-layer/text-layer.js [hunk 2]
```diff
@@ -266,63 +304,121 @@ export default class TextLayer extends CompositeLayer {
...
+              : data,
+            _dataDiff,
+            // Maintain the same background behavior as <=8.3. Remove in v9?
+            autoHighlight: false,
+            getBoundingRect: this.getBoundingRect.bind(this)
+          }
+        ),
+      new CharacgtersLayerClass(
+        {
+          sdf: fontSettings.sdf,
```
Comment (ibgreen) [inline modules/layers/src/text-layer/text-layer.js:371]: 400 lines, half of it props copying.

File: modules/layers/src/text-layer/text-background-layer/text-background-layer.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      .draw();
+  }
+
+  _getModel(gl) {
+    // a square that minimally cover the unit circle
+    const positions = [0, 0, 1, 0, 1, 1, 0, 1];
+
+    return new Model(
+      gl,
+      Object.assign(this.getShaders(), {
```
Comment (ibgreen) [inline modules/layers/src/text-layer/text-background-layer/text-background-layer.js:131]: Still using Object.assign?

**Review by ibgreen — APPROVED** [review]
Nice. Just an observation, these things require a lot of boiler plate code...

File: modules/layers/src/text-layer/text-layer.js (hunk 2 above)
Comment (Pessimistress) [reply modules/layers/src/text-layer/text-layer.js:371]: We decided to use explicit prop forwarding for all core layers a long time ago. `TileLayer` is the only exception and we are dealing with all sorts of issues.

File: modules/layers/src/text-layer/text-background-layer/text-background-layer.js (hunk 3 above)
Comment (Pessimistress) [reply modules/layers/src/text-layer/text-background-layer/text-background-layer.js:131]: I'll make a pass in a separate PR.

**Pull request merged**